providers/utils.go

```diff
@@ -0,0 +1,68 @@
+// Copyright 2025 OpenPubkey
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package providers
+
+import (
+	"crypto/rand"
+	"fmt"
+	"io"
+	"net"
+	"net/url"
+
+	httphelper "github.com/zitadel/oidc/v3/pkg/http"
+)
+
+// FindAvailablePort attempts to open a listener on localhost until it finds one or runs out of redirectURIs to try
+func FindAvailablePort(redirectURIs []string) (*url.URL, net.Listener, error) {
+	var ln net.Listener
+	var lnErr error
+	for _, v := range redirectURIs {
+		redirectURI, err := url.Parse(v)
+		if err != nil {
+			return nil, nil, fmt.Errorf("malformed redirectURI specified, redirectURI was %s", v)
+		}
+
+		lnStr := fmt.Sprintf("localhost:%s", redirectURI.Port())
```

Comment:
Hmm, do we really want to just take the port and change the host to localhost? Perhaps we should instead be skipping any redirect URIs that don't have a `localhost` host.

Reply:
Good point, added a check